File: UVA/401.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

char giveReverse(char c);
bool isPalindrome(char word[]);
bool isMirroredString(char word[]);
bool isMirroredPalindrome(char word[]);

int main(int argc, char const *argv[]) {
	char word[1000];
	while((scanf("%s", word)) != EOF) {
		if(isMirroredPalindrome(word))
            printf("%s%s", word, " -- is a mirrored palindrome.");
        else if(isMirroredString(word))
        	printf("%s%s", word, " -- is a mirrored string.");
        else if(isPalindrome(word))
        	printf("%s%s", word, " -- is a regular palindrome.");
        else
        	printf("%s%s", word, " -- is not a palindrome.");
        printf("\n\n");
	}

	return 0;
}

char giveReverse(char c) {
	if (c == 'A') return 'A';
	if (c == 'E') return '3';
    if (c == 'H') return 'H';
    if (c == 'I') return 'I';
    if (c == 'J') return 'L';
    if (c == 'L') return 'J';
    if (c == 'M') return 'M';
    if (c == 'O') return 'O';
    if (c == 'S') return '2';
    if (c == 'T') return 'T';
    if (c == 'U') return 'U';
    if (c == 'V') return 'V';
    if (c == 'W') return 'W';
    if (c == 'X') return 'X';
    if (c == 'Y') return 'Y';
    if (c == 'Z') return '5';
    if (c == '1') return '1';
    if (c == '2') return 'S';
    if (c == '3') return 'E';
    if (c == '5') return 'Z';
    if (c == '8') return '8';
    return '_';
}

bool isPalindrome(char word[]) {
	int i = 0, length = strlen(word);
	for(; i < length / 2; i++)
		if(word[i] != word[length - 1 - i])
			return false;
    return true;
}

bool isMirroredString(char word[]) {
	int i = 0, length = strlen(word);
	for(; i < length / 2; i++) 
		if(giveReverse(word[i]) == '_' || word[length - 1 - i] != giveReverse(word[i]))
			return false;
    return true;
}

bool isMirroredPalindrome(char word[]) {
    return isPalindrome(word) && isMirroredString(word);
}

File: UVA/-deprecated/401.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

class Main
{

    static final HashMap<Character, Character> hm = new HashMap<Character, Character>();
    static final boolean DEBUGGING = false;

    public static void main(String[] args) throws IOException {
        final BufferedReader in;
        hm.put('A', 'A');
        hm.put('E', '3');
        hm.put('H', 'H');
        hm.put('I', 'I');
        hm.put('J', 'L');
        hm.put('L', 'J');
        hm.put('M', 'M');
        hm.put('O', 'O');
        hm.put('S', '2');
        hm.put('T', 'T');
        hm.put('U', 'U');
        hm.put('V', 'V');
        hm.put('W', 'W');
        hm.put('X', 'X');
        hm.put('Y', 'Y');
        hm.put('Z', '5');
        hm.put('1', '1');
        hm.put('2', 'S');
        hm.put('3', 'E');
        hm.put('5', 'Z');
        hm.put('8', '8');
        if(!DEBUGGING)
            in = new BufferedReader(new InputStreamReader(System.in));
        else
            in = new BufferedReader(new FileReader("in.txt"));
        String line;
        while((line = in.readLine()) != null && !line.equals("")) {
            System.out.println(solve(line.trim()));
            System.out.println();
        }
    }

    static String solve(String word) {
        if(isMirroredPalindrome(word))
            return word + " -- is a mirrored palindrome.";
        if(isMirroredString(word))
            return word + " -- is a mirrored string.";
        if(isPalindrome(word))
            return word + " -- is a regular palindrome.";
        else
            return word + " -- is not a palindrome.";
    }

    static boolean isMirroredPalindrome(String word) {
        return isPalindrome(word) && isMirroredString(word);
    }

    static boolean isPalindrome(String string){
        int left = 0;
        int right = string.length() - 1;
        for(; left < right && string.charAt(left) == string.charAt(right); left++, right--);
        return left >= right;
    }

    static boolean isMirroredString(String string){
        for(int i = string.length() - 1; i >= 0; i--) {
            if(!hm.containsKey(string.charAt(i)) || 
                hm.get(string.charAt(i)) != string.charAt(string.length() - i - 1))
                    return false;
            }
        return true;
    }
}